File: app/src/main/java/com/examples/android/androidtrainingbzu/ScaleGestureDetectorActivity.java
package com.examples.android.androidtrainingbzu;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.widget.ImageView;
import android.widget.Toast;

import com.examples.android.androidtrainingbzu.Listeners.Android_Gesture_Detector;

public class ScaleGestureDetectorActivity extends AppCompatActivity {
    private ImageView imageView;
    private float scale = 1f;
    private ScaleGestureDetector detector;
    Android_Gesture_Detector android_gesture_detector;
    GestureDetector mGestureDetector;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_scale_gesture_detector);
        imageView=(ImageView)findViewById(R.id.img_scalimg);
        detector = new ScaleGestureDetector(this,new ScaleListener());
        android_gesture_detector=new Android_Gesture_Detector();
// Create a GestureDetector
        mGestureDetector = new GestureDetector(this, android_gesture_detector);
    }




    public boolean onTouchEvent(MotionEvent event) {
//  re-route the Touch Events to the ScaleListener class
        detector.onTouchEvent(event);
        mGestureDetector.onTouchEvent(event);
        return super.onTouchEvent(event);
    }

    private class ScaleListener extends ScaleGestureDetector.SimpleOnScaleGestureListener {


        float onScaleBegin = 0;
        float onScaleEnd = 0;

        @Override
        public boolean onScale(ScaleGestureDetector detector) {
            scale *= detector.getScaleFactor();
            imageView.setScaleX(scale);
            imageView.setScaleY(scale);
            return true;
        }

        @Override
        public boolean onScaleBegin(ScaleGestureDetector detector) {

            Toast.makeText(getApplicationContext(),"Scale Begin" ,Toast.LENGTH_SHORT).show();
            onScaleBegin = scale;

            return true;
        }

        @Override
        public void onScaleEnd(ScaleGestureDetector detector) {

            Toast.makeText(getApplicationContext(),"Scale Ended",Toast.LENGTH_SHORT).show();
            onScaleEnd = scale;

            if (onScaleEnd > onScaleBegin){
                Toast.makeText(getApplicationContext(),"Scaled Up by a factor of  " + String.valueOf( onScaleEnd / onScaleBegin ), Toast.LENGTH_SHORT  ).show();
            }

            if (onScaleEnd < onScaleBegin){
                Toast.makeText(getApplicationContext(),"Scaled Down by a factor of  " + String.valueOf( onScaleBegin / onScaleEnd ), Toast.LENGTH_SHORT  ).show();
            }

            super.onScaleEnd(detector);
        }
    }
}

File: app/src/main/java/com/examples/android/androidtrainingbzu/ScaleGestureDetectorActivity.kt
package com.examples.android.androidtrainingbzu

import android.os.Bundle
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.ScaleGestureDetector.SimpleOnScaleGestureListener
import android.view.View
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.examples.android.androidtrainingbzu.Listeners.Android_Gesture_Detector

class ScaleGestureDetectorActivity : AppCompatActivity() {
    private var imageView: ImageView? = null
    private var scale = 1f
    private var detector: ScaleGestureDetector? = null
    var android_gesture_detector: Android_Gesture_Detector? = null
    var mGestureDetector: GestureDetector? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_scale_gesture_detector)
        imageView = findViewById<View>(R.id.img_scalimg) as ImageView
        detector = ScaleGestureDetector(this, ScaleListener())
        android_gesture_detector = Android_Gesture_Detector()
        // Create a GestureDetector
        mGestureDetector = GestureDetector(this, android_gesture_detector)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
//  re-route the Touch Events to the ScaleListener class
        detector!!.onTouchEvent(event)
        mGestureDetector!!.onTouchEvent(event)
        return super.onTouchEvent(event)
    }

    inner class ScaleListener : SimpleOnScaleGestureListener() {
        var onScaleBegin = 0f
        var onScaleEnd = 0f
        override fun onScale(detector: ScaleGestureDetector): Boolean {
            scale *= detector.scaleFactor
            imageView!!.scaleX = scale
            imageView!!.scaleY = scale
            return true
        }

        override fun onScaleBegin(detector: ScaleGestureDetector): Boolean {
            Toast.makeText(applicationContext, "Scale Begin", Toast.LENGTH_SHORT).show()
            onScaleBegin = scale
            return true
        }

        override fun onScaleEnd(detector: ScaleGestureDetector) {
            Toast.makeText(applicationContext, "Scale Ended", Toast.LENGTH_SHORT).show()
            onScaleEnd = scale
            if (onScaleEnd > onScaleBegin) {
                Toast.makeText(applicationContext,
                    "Scaled Up by a factor of  " + (onScaleEnd / onScaleBegin).toString(),
                    Toast.LENGTH_SHORT).show()
            }
            if (onScaleEnd < onScaleBegin) {
                Toast.makeText(applicationContext,
                    "Scaled Down by a factor of  " + (onScaleBegin / onScaleEnd).toString(),
                    Toast.LENGTH_SHORT).show()
            }
            super.onScaleEnd(detector)
        }
    }
}